src/muxer/yamux/yamuxed_connection.cpp

```diff
@@ -3,89 +3,107 @@
-             "in order to create a new one";
-    case ErrorType::FORBIDDEN_CALL:
-      return "forbidden method was invoked";
-    case ErrorType::OTHER_SIDE_ERROR:
-      return "error happened on other side's behalf";
-    case ErrorType::INTERNAL_ERROR:
-      return "internal error happened";
-    case ErrorType::CLOSED_BY_PEER:
-      return "connection closed by peer";
-  }
-  return "unknown";
-}
+#define TRACE_ENABLED 1
+#include <libp2p/common/trace.hpp>
 
 namespace libp2p::connection {
+
+  namespace {
+    auto log() {
+      static auto logger = common::createLogger("yx-conn");
+      return logger.get();
+    }
+
+    inline size_t size(const gsl::span<uint8_t> &span) {
+      return static_cast<size_t>(span.size());
+    }
+
+    inline std::tuple<gsl::span<uint8_t>, gsl::span<uint8_t>> split(
+        gsl::span<uint8_t> span, size_t n) {
+      return {span.first(n), span.subspan(n)};
+    }
+
+    inline bool isOutbound(uint32_t our_stream_id, uint32_t their_stream_id) {
+      // streams id oddness and evenness, depends on connection direction,
+      // outbound or inbound, resp.
+      return ((our_stream_id ^ their_stream_id) & 1) == 0;
+    }
+
+  }  // namespace
+
   YamuxedConnection::YamuxedConnection(
       std::shared_ptr<SecureConnection> connection,
       muxer::MuxedConnectionConfig config)
-      : header_buffer_(YamuxFrame::kHeaderLength, 0),
-        data_buffer_(config.maximum_window_size, 0),
-        connection_{std::move(connection)},
-        config_{config} {
-    // client uses odd numbers, server - even
-    last_created_stream_id_ = connection_->isInitiator() ? 1 : 2;
+      : config_(config),
+        connection_(std::move(connection)),
+        raw_read_buffer_(std::make_shared<Buffer>()),
+        reading_state_(
+            [this](boost::optional<YamuxFrame> header) {
+              return processHeader(std::move(header));
+            },
+            [this](gsl::span<uint8_t> segment, StreamId stream_id, bool rst,
+                   bool fin) {
+              if (!segment.empty()) {
+                if (!processData(segment, stream_id)) {
+                  return false;
+                }
+              }
+              if (rst) {
+                return processRst(stream_id);
+              }
+              if (fin) {
+                return processFin(stream_id);
+              }
+              return true;
+            }) {
+    assert(connection_);
+    assert(config_.maximum_streams > 0);
+    assert(config_.maximum_window_size >= YamuxFrame::kInitialWindowSize);
+
+    raw_read_buffer_->resize(YamuxFrame::kInitialWindowSize + 4096);
+    new_stream_id_ = (connection_->isInitiator() ? 1 : 2);
   }
 
   void YamuxedConnection::start() {
-    BOOST_ASSERT_MSG(!started_,
-                     "YamuxedConnection already started (double start)");
+    if (started_) {
+      log()->critical("already started (double start)");
```

Comment:
It is really need to have 'critical' level? As I know, this level using for situation, when application cannot continue to work anymore.